[HttpFoundation] Similar locale selection

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | none
| License       | MIT

Allow the nearest locale to be selected instead the default one.
Ping @welcoMattic 

I noted a non-optimized locale selection. Let say I have the following controller and root route:

```php
#[Route('/', name: 'app_root', methods: [Request::METHOD_GET])]
public function indexNoLocale(Request $request): Response
{
    $locale = $request->getPreferredLanguage($this->supportedLocales) ?? $this->defaultLocale;

    return $this->redirectToRoute('app_homepage', [
        '_locale' => $locale,
    ], Response::HTTP_SEE_OTHER);
}
```

And the following parameters:
* `$this->supportedLocales` = `['fr_FR', 'en_US']`
* `$this->defaultLocale` = `'en_US'`

When a user arrives on this page with the header `accept-language: ja-JP,fr_CA;q=0.7,fr;q=0.5`, the default locale `en_US` is returned.
In this situation, I would expect to use `fr_FR` as the browser indicates French is a possible choice (`fr_CA` and `fr`).

With this change, and if no exact match is found, a similar language is selected. In this example, `fr_FR` is acceptable and then returned.

Nyholm reviewed (commented):
Thank you. It looks cool.

Can you also test these: 

```
ja-JP,fr;q=0.5
ja-JP,fr_CA;q=0.7
```

Both of them should go to `fr_FR`, right? 

----

How about these?

```
ja-JP,fr_CA;q=0.7,fr;q=0.5,en_US;q=0.3
ja-JP,fr;q=0.5,en_US;q=0.3
ja-JP,fr_CA;q=0.7,en_US;q=0.3
ja-JP,fr_CA;q=0.7,en;q=0.5
```

welcoMattic commented:
Maybe for another PR, but shouldn't we add support for `q=` parameter? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#directives

Spomky commented:
Hi,

I updated the PR and changed the comparison logic:
* If an exact match exists, it is returned
* Else if a similar locale exists, the corresponding language is returned
* If nothing is acceptable, the first supported locale is returned

Please note that the `q` parameter is already used. The [test with a reversed order priority](https://github.com/symfony/symfony/pull/52986/files#diff-a6e313421b40e37ad1c1a23ad66131c48790cb435dc1d040ac33a4800e243780R1553) returns the expected language.

In addition, the returned locale value is always `xx_XX` to make it consistant with the value used internally. 
Also, it accepts locales such as `hi_Latn_IN` to be compared.

yceruto commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1609,6 +1609,7 @@ public function getPreferredFormat(?string $default = 'html'): ?string
     public function getPreferredLanguage(array $locales = null): ?string
     {
         $preferredLanguages = $this->getLanguages();
+        $locales = array_map([$this, 'formatLocale'], $locales ?? []);
```
Since PHP 8.1, the project preference is to use first-class callable syntax instead of the array syntax.

```suggestion
        $locales = array_map($this->formatLocale(...), $locales ?? []);
```

yceruto commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1662,39 @@ public function getLanguages(): array
 
+    private static function formatLocale(string $locale): string
+    {
+        if (str_contains($locale, '-')) {
```
Now that you have moved this code to a new method, let's refactor it a bit by returning early if `!str_contains($locale, '-')`, thus removing one nested level.

Spomky replied:
You are right, I will update this PR and remove a level of nesting.

stloyd commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1662,40 @@ public function getLanguages(): array
+            // of any listed language, which can be registered with the
+            // i-prefix, such as i-cherokee
+            if (\count($codes) > 1) {
+                $locale = $codes[1];
```
```suggestion
                return $codes[1];
```

Spomky replied:
Hi @stloyd,

Many thanks for the suggestion.
I will not commit it as I am working on this part that needs other modifications.
I noticed issues when variant are present in the locale e.g. zh-hans-tw (simplified Chinese from Taiwan) or when components/extensions follow the region (`pt-BR-u-ca-gregory-nu-latn` for Brazilian Portuguese with use of the Gregorian calendar and Latin numbering).
This function should return `zh_Hans_TW` or `pt_BR`, but it does not and an exact match may be missed because of this.
For `zh_Hans_TW`, it could fallback to `zh_TW` and `zh` but this is not the case at the moment.

@welcoMattic I think we could add a class to handle these locale parts in the Translation component. This will help the Request object handle the languages and regions parts correctly to find the most appropriate locale. WDYT?
See https://en.wikipedia.org/wiki/IETF_language_tag or https://andreipfeiffer.dev/slides/i18n-and-l10n.pdf

Something like as follows

```php
final class LanguageTag
{
    public string $language;
    public ?string $script = null;
    public ?string $region = null;
    public array $variants = [];
    public array $extensions = [];

    public function __construct(string $languageTag)
    {
        $components = explode('-', strtolower($languageTag));
        $this->language = array_shift($components);

        foreach ($components as $component) {
            if (strlen($component) === 4) {
                $this->script = ucfirst($component);
            } elseif (strlen($component) === 2) {
                $this->region = mb_strtoupper($component);
            }
            // Variants and extensions to handle
        }
    }
}
```

welcoMattic replied:
I'm wondering the right place for this class, Translation or Intl? @Nyholm @OskarStark any thoughts?

Spomky replied:
Note that for this PR we only need to capture the main parts.

Spomky replied:
Hi,

I completely reviewed the method and it now able to differentiate locales with language/script/region.
Also, I added some tests to make sure the selection is more accurate.

OskarStark replied:
> I'm wondering the right place for this class, Translation or Intl? @Nyholm @OskarStark any thoughts?

My feeling is about intl, but not sure

cc @stof

Spomky replied:
I will start working on this soon. I'll create it in the Intl component and see if it needs to be moved elsewhere as it progresses.

smnandre replied:
That is what intl provide (the PHP extension) with the [Locale](https://www.php.net/manual/en/class.locale.php) classes.

symfony/intl is kinda polyfill for those without intl extension. You may want to look at the int-contracts, but i really do not think symfony/intl should be the place for that.

As nowadays it contains massive amount of data... and finally not much class / helpers.

Spomky replied:
OK thanks. I will reconsider this.
For this PR, there is no need to depend on other components.

smnandre commented:
In your exemple, what would happen if the default locale were defined as « fr, en » ?

Spomky commented:
> In your exemple, what would happen if the default locale were defined as « fr, en » ?

You mean if the supported locales are `['fr', 'en']` instead of `['fr-FR', 'en-US']`? In this case it will match ass `accept-language: ja-JP,fr_CA;q=0.7,fr;q=0.5` contains the exact match `fr`.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1654,85 @@ public function getLanguages(): array
+    }
+
+    /**
+     * This method returns an array of all possible combinations of the language components.
```
```suggestion
     * Returns an array of all possible combinations of the language components.
     *
```

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1654,85 @@ public function getLanguages(): array
     }
 
+    /**
+     * This method strips the locale to only keep the canonicalized language value.
```
blank line after (to separate the title from the description)

```suggestion
     * Strips the locale to only keep the canonicalized language value.
     *
```

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1654,85 @@ public function getLanguages(): array
+    {
+        $locale = str_replace('_', '-', strtolower($locale));
+        $pattern = '/^([a-zA-Z]{2,3}|i-[a-zA-Z]{5,})(?:-([a-zA-Z]{4}))?(?:-([a-zA-Z]{2}))?(?:-(.+))?$/';
+        if (1 !== preg_match($pattern, $locale, $matches)) {
```
```suggestion
        if (!preg_match($pattern, $locale, $matches)) {
```

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1618,20 +1619,24 @@ public function getPreferredLanguage(array $locales = null): ?string
 
-        $extendedPreferredLanguages = [];
+        $matches = array_values(array_intersect($preferredLanguages, $locales));
+        if ($matches) {
```
```suggestion
        if ($matches = array_values(array_intersect($preferredLanguages, $locales))) {
```

nicolas-grekas reviewed (commented):
Here's some nitpicking. What's the state of this PR? Any votes from previous reviewers?

Spomky commented:
Many thanks for the last suggestions.
The PR is now rebased, squashed and ready for other reviewers comments

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1618,20 +1619,23 @@ public function getPreferredLanguage(?array $locales = null): ?string
         }
 
-        $extendedPreferredLanguages = [];
+        if ($matches = array_values(array_intersect($preferredLanguages, $locales))) {
```
we'd better remove `array_values` and `return reset($matches);` on the next line

Spomky replied:
Yes. Done.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1618,20 +1619,23 @@ public function getPreferredLanguage(?array $locales = null): ?string
+        if ($matches = array_values(array_intersect($preferredLanguages, $locales))) {
+            return $matches[0];
+        }
+        $combinations = [];
```
`$combinations = array_merge(...array_map($this->getLanguageCombinations(...), $peferredLanguages));`

Spomky replied:
The shorter, the better. Done

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1653,87 @@ public function getLanguages(): array
+
+    /**
+     * This method returns an array with the language components of the locale.
+     * For example:
```
missing blank line before this line

Spomky replied:
Blak line added.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1653,87 @@ public function getLanguages(): array
-                if ('i' === $codes[0]) {
-                    // Language not listed in ISO 639 that are not variants
-                    // of any listed language, which can be registered with the
-                    // i-prefix, such as i-cherokee
```
can we keep this comment somewhere?

Spomky replied:
I restored this comment on line 1643 to 1645

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1653,87 @@ public function getLanguages(): array
+     * @see https://wikipedia.org/wiki/IETF_language_tag
+     * @see https://datatracker.ietf.org/doc/html/rfc5646
+     *
+     * @return array{language: string, script: string|null, region: string|null}
```
```suggestion
     * @return array{string, string|null, string|null}
```

Spomky replied:
Done

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1618,20 +1619,23 @@ public function getPreferredLanguage(?array $locales = null): ?string
+            $combinations = array_merge($combinations, $this->getLanguageCombinations($language));
+        }
+
+        foreach (array_unique($combinations) as $combination) {
```
let's skip array_unique

Spomky replied:
`array_unique` is now gone

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1649,32 +1653,87 @@ public function getLanguages(): array
+     * - "fr_Latn_FR"
+     * - "fr_Latn"
+     * - "fr_FR"
+     * - "fr".
```
```suggestion
     * - "fr"
```

Spomky replied:
DOne

nicolas-grekas reviewed (commented):
That's an improvement, for 7.1 ;)

smnandre commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1609,6 +1609,7 @@ public function getPreferredFormat(?string $default = 'html'): ?string
     public function getPreferredLanguage(?array $locales = null): ?string
     {
         $preferredLanguages = $this->getLanguages();
+        $locales = array_map($this->formatLocale(...), $locales ?? []);
```
Should be moved line 1621 maybe ? (micro-picking :) )

Spomky replied:
Indeed. It moved this line

dunglas commented:
It looks like this isn't a good practice according to the spec: https://www.rfc-editor.org/rfc/rfc9110#field.accept-language

If `fr_BE` is requested, `fr_FR` shouldn't be sent. Instead, the browser should include both `fr_BE` and `fr` in `Accept-Language`.

stof commented:
PHP already has a locale matching algorithm in ext-intl: https://www.php.net/manual/fr/locale.lookup.php (or maybe some of the other methods there)

Would this algorithm help for this PR ? If yes, the proper way to solve it might be to implement them in the polyfill in https://github.com/symfony/polyfill/blob/1.x/src/Intl/Icu/Locale.php (where it currently throw an exception saying it is not implemented) and to use it.

nicolas-grekas reviewed (approved):
LGTM thanks

> PHP already has a locale matching algorithm in ext-intl [...] the proper way to solve it might be to implement them in the polyfill

Good idea! I suggest merging this PR, and switch to the polyfill once we have it. Nothing is public here so the switch should be easy.

OskarStark commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1567,32 +1568,91 @@ public function getLanguages(): array
+     * - language_Script_REGION: "fr_Latn_FR", "zh_Hans_TW"
+     * - language_Script: "fr_Latn", "zh_Hans"
+     * - language_REGION: "fr_FR", "zh_TW"
+     * - language: "fr", "zh".
```
```suggestion
     * - language: "fr", "zh"
```

OskarStark commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1567,32 +1568,91 @@ public function getLanguages(): array
+     * For example:
+     * - If the locale is "fr_Latn_FR", this method will return "fr", "Latn", "FR"
+     * - If the locale is "fr_FR", this method will return "fr", null, "FR"
+     * - If the locale is "zh_Hans", this method will return "zh", "Hans", null.
```
```suggestion
     * - If the locale is "zh_Hans", this method will return "zh", "Hans", null
```

OskarStark commented on src/Symfony/Component/HttpFoundation/Tests/RequestTest.php:
```diff
@@ -2199,7 +2213,7 @@ public function testFactory()
     public function testFactoryCallable()
     {
-        $requestFactory = new class {
+        $requestFactory = new class() {
```
not sure about our code style here, with `()` or without

Spomky replied:
fabbot patched it and added `()`

smnandre commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1567,32 +1568,91 @@ public function getLanguages(): array
+    /**
+     * Returns an array of all possible combinations of the language components.
+     *
+     * For example, if the locale is "fr_Latn_FR", this method will return :
```
```suggestion
     * For instance, if the locale is "fr_Latn_FR", this method will return:
```

Spomky committed:
Allow the nearest locale to be selected instead the default one.

fabpot commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1567,32 +1568,91 @@ public function getLanguages(): array
+    }
+
+    /**
+     * This method returns an array with the language components of the locale.
```
```suggestion
     * Returns an array with the language components of the locale.
```

fabpot commented:
Thank you @Spomky.
